adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java

```diff
@@ -279,4 +282,22 @@ private void mapDownstreamMessageRequest(
                 resultHandler.handle(result.future());
             });
     }
+
+    private Throwable mapException(final AsyncResult<HttpResponse<Buffer>> httpResponseAsyncResult) {
+        final Optional<HttpResponse<Buffer>> httpResponse = Optional.ofNullable(httpResponseAsyncResult.result());
+        final int statusCode = httpResponse.map(HttpResponse::statusCode).orElse(HttpURLConnection.HTTP_INTERNAL_ERROR);
+        if (statusCode >= 400 && statusCode < 500) {
+            return new ClientErrorException(statusCode, httpResponseAsyncResult.cause());
+        }
+        return new ServerErrorException(statusCode, httpResponseAsyncResult.cause());
+    }
+
+    private Throwable mapException(final AsyncResult<HttpResponse<Buffer>> httpResponseAsyncResult, final String message) {
+        final Optional<HttpResponse<Buffer>> httpResponse = Optional.ofNullable(httpResponseAsyncResult.result());
+        final int statusCode = httpResponse.map(HttpResponse::statusCode).orElse(HttpURLConnection.HTTP_INTERNAL_ERROR);
+        if (statusCode >= 400 && statusCode < 500) {
+            return new ClientErrorException(statusCode, message);
+        }
+        return new ServerErrorException(statusCode, message);
+    }
```

Comment:
can we reduce the code duplication by e.g. having just one method which accepts an `Optional<String> message`?

Reply:
you might also want to consider using `org.eclipse.hono.client.util.StatusCodeMapper.from()` to create a context specific `ServiceInvocationException` ...